### PR TITLE
use sacremoses normalizer, ensure pretranslate.src.json and pretranslate.trg.json use same directory

### PR DESCRIPTION
**Pull request by mshannon-sil**
I added sacremoses normalization before training and before pretranslation. I also made sure that `pretranslate.trg.json` makes use of the `shared_file_uri` to ensure it ends up in the same directory as `pretranslate.src.json`. There's also a small fix I made to `launch.json` because I was getting an error that the build options were not valid json.

- - -
This change is [Reviewable](https://reviewable.io/reviews/sillsdev/machine.py/96)

### CODE REVIEW

**Review by ddaspit — CHANGES_REQUESTED** [review]
> Reviewed 4 of 4 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/sillsdev/machine.py/96)* status: all files reviewed, 4 unresolved discussions (waiting on @johnml1135 and @mshannon-sil)

___
*[`.vscode/launch.json` line 34 at r1](https://reviewable.io/reviews/sillsdev/machine.py/96#-NnpSNJd-XWKJUz2RdAS:-NnpSNJd-XWKJUz2RdAT:b-nxofsl) ([raw file](https://github.com/sillsdev/machine.py/blob/add79fb0cbbb9147ebb21ec9d1c59dc760c0ae6e/.vscode/launch.json#L34)):*
> ```JSON
>           "eng_Latn",
>           "--build-options",
>           "{\"train_params\":{\"max_steps\":10}}"
> ```

I think the extra slashes are needed on Windows.
___
*[`machine/jobs/shared_file_service.py` line 67 at r1](https://reviewable.io/reviews/sillsdev/machine.py/96#-NnpbydQ8DpCKx7J2Tzw:-NnpbydQ8DpCKx7J2Tzx:bvdmb19) ([raw file](https://github.com/sillsdev/machine.py/blob/add79fb0cbbb9147ebb21ec9d1c59dc760c0ae6e/machine/jobs/shared_file_service.py#L67)):*
> ```Python
>     @contextmanager
>     def open_target_pretranslation_writer(self) -> Iterator[PretranslationWriter]:
>         def remove_prefix(text: str, prefix: str) -> str:
> ```

The original code works correctly on my dev machine (Windows). This fix breaks the code. What exactly is the issue that you are seeing?
___
*[`machine/translation/huggingface/hugging_face_nmt_engine.py` line 173 at r1](https://reviewable.io/reviews/sillsdev/machine.py/96#-NnpTb4w4dXuaB2djbWU:-NnpTb4w4dXuaB2djbWV:b-dnwxy3) ([raw file](https://github.com/sillsdev/machine.py/blob/add79fb0cbbb9147ebb21ec9d1c59dc760c0ae6e/machine/translation/huggingface/hugging_face_nmt_engine.py#L173)):*
> ```Python
>         if isinstance(self.tokenizer, (NllbTokenizer, NllbTokenizerFast)):
>             mpn = MosesPunctNormalizer()
>             mpn.substitutions = [(re.compile(r), sub) for r, sub in mpn.substitutions]
> ```

This is an "expensive" operation. It would be better if this is performed only once. In the constructor of the `HuggingFaceNmtEngine` class you can create the `MosesPunctNormalizer` and assign it to a field. It would also be good to perform the normalization when the `sentences` list is first constructed, so you aren't iterating and constructing a list twice.
___
*[`machine/translation/huggingface/hugging_face_nmt_model_trainer.py` line 311 at r1](https://reviewable.io/reviews/sillsdev/machine.py/96#-NnpVSEE1abfm8nSBjLN:-NnpVSEE1abfm8nSBjLO:b-dnwxy3) ([raw file](https://github.com/sillsdev/machine.py/blob/add79fb0cbbb9147ebb21ec9d1c59dc760c0ae6e/machine/translation/huggingface/hugging_face_nmt_model_trainer.py#L311)):*
> ```Python
>             if isinstance(tokenizer, (NllbTokenizer, NllbTokenizerFast)):
>                 mpn = MosesPunctNormalizer()
>                 mpn.substitutions = [(re.compile(r), sub) for r, sub in mpn.substitutions]
> ```

This is an "expensive" operation. It would be better if this is performed only once. In the constructor of the `HuggingFaceNmtEngine` class you can create the `MosesPunctNormalizer` and assign it to a field. It would also be good to perform the normalization when the `sentences` list is first constructed, so you aren't iterating and constructing a list twice.

**Review by ddaspit — COMMENTED** [review]
> Reviewed 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/sillsdev/machine.py/96)* status: all files reviewed, 3 unresolved discussions (waiting on @johnml1135 and @mshannon-sil)

**Review by mshannon-sil — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/sillsdev/machine.py/96)* status: 2 of 4 files reviewed, 3 unresolved discussions (waiting on @ddaspit and @johnml1135)

___
*[`machine/translation/huggingface/hugging_face_nmt_engine.py` line 173 at r1](https://reviewable.io/reviews/sillsdev/machine.py/96#-NnpTb4w4dXuaB2djbWU:-Nnus3rxC58b99Fa_seP:b7h4ku0) ([raw file](https://github.com/sillsdev/machine.py/blob/add79fb0cbbb9147ebb21ec9d1c59dc760c0ae6e/machine/translation/huggingface/hugging_face_nmt_engine.py#L173)):*
<details><summary><i>Previously, ddaspit (Damien Daspit) wrote…</i></summary><blockquote>

This is an "expensive" operation. It would be better if this is performed only once. In the constructor of the `HuggingFaceNmtEngine` class you can create the `MosesPunctNormalizer` and assign it to a field. It would also be good to perform the normalization when the `sentences` list is first constructed, so you aren't iterating and constructing a list twice.
</blockquote></details>

Made some updates to address this, and I'm ready for another review on it.

**Review by ddaspit — COMMENTED** [review]
> Reviewed 2 of 2 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/sillsdev/machine.py/96)* status: all files reviewed, 2 unresolved discussions (waiting on @johnml1135 and @mshannon-sil)

___
*[`machine/jobs/shared_file_service.py` line 67 at r1](https://reviewable.io/reviews/sillsdev/machine.py/96#-NnpbydQ8DpCKx7J2Tzw:-Nny8jz31TktasQouqv4:b-jr13f9) ([raw file](https://github.com/sillsdev/machine.py/blob/add79fb0cbbb9147ebb21ec9d1c59dc760c0ae6e/machine/jobs/shared_file_service.py#L67)):*
<details><summary><i>Previously, ddaspit (Damien Daspit) wrote…</i></summary><blockquote>

The original code works correctly on my dev machine (Windows). This fix breaks the code. What exactly is the issue that you are seeing?
</blockquote></details>

I think it will make it easier to fix this issue if we refactor the `settings.yaml` a bit:
- Update `shared_file_uri` to just the base URI and move it to `default`, i.e. `s3://aqua-ml-data/`
- Add a new setting called `shared_file_folder`, set it to `production` in `default`, `dev` in `development`, and `ext-qa` in `staging`.
___
*[`machine/translation/huggingface/hugging_face_nmt_model_trainer.py` line 174 at r3](https://reviewable.io/reviews/sillsdev/machine.py/96#-Nny8X1p3hHSVtH1fmPZ:-Nny8X1p3hHSVtH1fmP_:bpnay29) ([raw file](https://github.com/sillsdev/machine.py/blob/8629f3f0b77ebb25705b0595f329c48316eb5d09/machine/translation/huggingface/hugging_face_nmt_model_trainer.py#L174)):*
> ```Python
>                 for ex in train_dataset["translation"]:
>                     charset = charset | set(ex[lang_code])
>             charset = {self._mpn.normalize(char) for char in charset}
> ```

This should only be done if the NLLB model is used.

**Review by mshannon-sil — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/sillsdev/machine.py/96)* status: 2 of 6 files reviewed, 2 unresolved discussions (waiting on @ddaspit and @johnml1135)

___
*[`machine/jobs/shared_file_service.py` line 67 at r1](https://reviewable.io/reviews/sillsdev/machine.py/96#-NnpbydQ8DpCKx7J2Tzw:-NnzQUk236iZrWeiCmL3:b-vu3h3q) ([raw file](https://github.com/sillsdev/machine.py/blob/add79fb0cbbb9147ebb21ec9d1c59dc760c0ae6e/machine/jobs/shared_file_service.py#L67)):*
<details><summary><i>Previously, ddaspit (Damien Daspit) wrote…</i></summary><blockquote>

I think it will make it easier to fix this issue if we refactor the `settings.yaml` a bit:
- Update `shared_file_uri` to just the base URI and move it to `default`, i.e. `s3://aqua-ml-data/`
- Add a new setting called `shared_file_folder`, set it to `production` in `default`, `dev` in `development`, and `ext-qa` in `staging`.
</blockquote></details>

This makes sense to me. I just pushed a change refactoring settings.yaml and adjusting the code to use those variables. I think that we'll also have to change how this works in Serval/Machine then, since Serval's deployment files currently only include a `SharedFileLocation` which have the uri and folder combined. That's used to set the secret `SharedFile__Uri`, which I think is used to set the value of`shared_file_uri` in machine.py, although I didn't spot where that was happening in the code. Should I open up an issue in Serval to add `shared_file_uri` and `shared_file_folder` there too?
___
*[`machine/translation/huggingface/hugging_face_nmt_model_trainer.py` line 174 at r3](https://reviewable.io/reviews/sillsdev/machine.py/96#-Nny8X1p3hHSVtH1fmPZ:-NnzXQTeE_NoAQ4kuqW1:b-896fix) ([raw file](https://github.com/sillsdev/machine.py/blob/8629f3f0b77ebb25705b0595f329c48316eb5d09/machine/translation/huggingface/hugging_face_nmt_model_trainer.py#L174)):*
<details><summary><i>Previously, ddaspit (Damien Daspit) wrote…</i></summary><blockquote>

This should only be done if the NLLB model is used.
</blockquote></details>

Done.

**Review by ddaspit — APPROVED** [review]
:lgtm:

> Reviewed 4 of 4 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/sillsdev/machine.py/96#-:-NnzbJ3M6yJ2KPhH4EMs:bnfp4nl)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved (waiting on @johnml1135)

___
*[`machine/jobs/shared_file_service.py` line 67 at r1](https://reviewable.io/reviews/sillsdev/machine.py/96#-NnpbydQ8DpCKx7J2Tzw:-NnzawHn9yLhksKNCmHx:b7557od) ([raw file](https://github.com/sillsdev/machine.py/blob/add79fb0cbbb9147ebb21ec9d1c59dc760c0ae6e/machine/jobs/shared_file_service.py#L67)):*
<details><summary><i>Previously, mshannon-sil wrote…</i></summary><blockquote>

This makes sense to me. I just pushed a change refactoring settings.yaml and adjusting the code to use those variables. I think that we'll also have to change how this works in Serval/Machine then, since Serval's deployment files currently only include a `SharedFileLocation` which have the uri and folder combined. That's used to set the secret `SharedFile__Uri`, which I think is used to set the value of`shared_file_uri` in machine.py, although I didn't spot where that was happening in the code. Should I open up an issue in Serval to add `shared_file_uri` and `shared_file_folder` there too?
</blockquote></details>

We need to update Machine to pass the `shared_file_uri` and `shared_file_folder` separately when the ClearML job is started.

**Comment by johnml1135** [comment]
This is to fix https://github.com/sillsdev/machine.py/issues/94.
